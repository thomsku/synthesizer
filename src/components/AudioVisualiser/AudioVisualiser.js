import React, {Component} from "react";
import PropTypes from "prop-types";

class AudioVisualiser extends Component {
  constructor(props) {
    super(props);
    this.canvas = React.createRef();
  }

  componentDidUpdate() {
    this.draw();
  }

  draw() {
    const canvas = this.canvas.current;
    const width = canvas.width;
    const height = canvas.height;
    const context = canvas.getContext("2d");
    const scaling = height / 256;
    let risingEdge = 0;
    const edgeThreshold = 5;

    context.fillStyle = "rgba(40, 56, 68, 0.21)";
    context.clearRect(0, 0, width, height);

    context.lineWidth = 2;
    context.strokeStyle = "#fff";
    context.beginPath();

    // No buffer overrun protection
    while (this.props.audioData[risingEdge++] - 128 > 0 && risingEdge <= width);
    if (risingEdge >= width) risingEdge = 0;

    while (this.props.audioData[risingEdge++] - 128 < edgeThreshold && risingEdge <= width);
    if (risingEdge >= width) risingEdge = 0;

    for (let x = risingEdge; x < this.props.audioData.length && x - risingEdge < width; x++) {
      context.lineTo(x - risingEdge, height - this.props.audioData[x] * scaling);
    }

    context.stroke();
  }

  render() {
    return <canvas width="800" height="256" ref={this.canvas} />;
  }
}

AudioVisualiser.propTypes = {
  audioData: PropTypes.object.isRequired,
};

export default AudioVisualiser;
