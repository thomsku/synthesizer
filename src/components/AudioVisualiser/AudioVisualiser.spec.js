import React from "react";
import {shallow, mount} from "enzyme";
import AudioVisualiser from "../AudioVisualiser/AudioVisualiser";

describe("AudioVisualiser", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<AudioVisualiser audioData={{}}/>));

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("should render a <canvas />", () => {
    expect(wrapper.find("canvas")).toHaveLength(1);
  });
});

describe("draw", () => {
  let wrapper;
  let audioData;

  window.HTMLCanvasElement.prototype.getContext = () => {
    return {
      width: 0,
      beginPath: jest.fn(),
      clearRect: jest.fn(),
      lineTo: jest.fn(),
      moveTo: jest.fn(),
      stroke: jest.fn(),
    };
  };

  beforeEach(() => {
    wrapper = mount(<AudioVisualiser audioData={{}}/>);
    audioData = new Uint8Array(1024);
  });

  it("is executed when the audioData prop is updated", () => {
    audioData.fill(128);
    const spy = jest.spyOn(wrapper.instance(), "draw");
    wrapper.setProps({audioData: audioData});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("handles negative audioData values", () => {
    audioData.fill(-1);
    const spy = jest.spyOn(wrapper.instance(), "draw");
    wrapper.setProps({audioData: audioData});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("handles too high audioData values", () => {
    audioData.fill(12000);
    const spy = jest.spyOn(wrapper.instance(), "draw");
    wrapper.setProps({audioData: audioData});
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
