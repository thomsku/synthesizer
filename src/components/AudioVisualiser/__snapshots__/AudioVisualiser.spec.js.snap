// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioVisualiser should render correctly 1`] = `
<canvas
  height="256"
  width="800"
/>
`;
